packages/fast-components/src/form-associated/index.ts

```diff
@@ -0,0 +1,263 @@
+
+/**
+ * Disable member ordering to keep property callbacks
+ * grouped with property declaration
+ */
+/* tslint:disable:member-ordering */
+export abstract class FormAssociated<
+    T extends HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement
+> extends FastElement {
+    /**
+     * Must evaluate to true to enable elementInternals.
+     * Feature detects API support and resolve respectively
+     */
+    public static get formAssociated(): boolean {
+        return "ElementInternals" in window;
+    }
+
+    /**
+     * Returns the validty state of the element
+     */
+    public get validity(): ValidityState {
+        return FormAssociated.formAssociated
+            ? this.elementInternals.validity
+            : this.proxy.validity;
+    }
+
+    /**
+     * Retrieve a reference to the associated form.
+     * Returns null if not associated to any form.
+     */
+    public get form(): HTMLFormElement | null {
+        return FormAssociated.formAssociated
+            ? this.elementInternals.form
+            : this.proxy.form;
+    }
+
+    /**
+     * Retrieve the localized validation message,
+     * or custom validation message if set.
+     */
+    public get validationMessage(): string {
+        return FormAssociated.formAssociated
+            ? this.elementInternals.validationMessage
+            : this.proxy.validationMessage;
+    }
+
+    /**
+     * Whether the element will be validated when the
+     * form is submitted
+     */
+    public get willValidate(): boolean {
+        return FormAssociated.formAssociated
+            ? this.elementInternals.willValidate
+            : this.proxy.willValidate;
+    }
+
+    /**
+     * A reference to all associated label elements
+     */
+    public get labels(): Node[] {
+        if (FormAssociated.formAssociated) {
+            return Array.from(this.elementInternals.labels);
+        } else if (
+            this.proxy instanceof HTMLElement &&
+            this.proxy.ownerDocument &&
+            this.id
+        ) {
+            // Labels associated by wraping the element: <label><custom-element></custom-element></label>
+            const parentLabels = this.proxy.labels;
+            // Labels associated using the `for` attribute
+            const forLabels = Array.from(
+                this.proxy.ownerDocument.querySelectorAll(`[for='${this.id}']`)
+            );
+
+            return !!parentLabels
+                ? forLabels.concat(Array.from(parentLabels))
+                : forLabels;
+        } else {
+            return [];
+        }
+    }
+
+    /**
+     * The value of the element to be associated with the form
+     */
+    @attr
+    public value: string = "";
+
+    /**
+     * Whether the element should be autofocused
+     */
+    @attr
+    public autofocus: boolean; // Map to proxy element
+
+    @attr
+    public disabled: boolean = false;
+    private disabledChanged(): void {
+        if (this.proxy instanceof HTMLElement) {

```

Comment:
In what case will proxy not bee an HTMLElement? Is this the case where we are using element internals instead of the proxy?

Reply:
No, its during initialization. I think it is a side effect of the `@attr` decorator - none of the member properties are available the first time this cb is called.